Site Setup: Use flow entry point to manage back button edgy cases

gabrielcaires opened the pull request:
Related to:
* https://github.com/Automattic/wp-calypso/issues/100895
* https://github.com/Automattic/jetpack/pull/42718

> [!WARNING]
> This current solution doesn't ensure an alignment between our UI back button and the browser back button.  Make them aligned is part of a refactor initiated on https://github.com/Automattic/wp-calypso/pull/101550.  

## Proposed Changes
* Rename the flow state from `migration`>`entryPoint` to `flow`>`entryPoint` because now it will be used outside the migration flow context. 
* Update back logic make the user return from IMPORT_OR_MIGRATE to {SITE_}/import.php when the ref= 'wp-admin-importers-list' 


The overall idea is the [PR](https://github.com/Automattic/jetpack/pull/42718) will introduce a specialized `ref=wp-admin-importers-list`, which will be stored on the new `site-setup` flow state to be reused when the user advance in on `site-setup` steps and use the back button.

We must also repass this ref query param to the site migration flow to ensure the user can properly return to the /import.php page.


![image](https://github.com/user-attachments/assets/93402536-26a8-4f7c-a3f1-1991cbabd065)


## Why are these changes being made?
* As part of the Code Blue initiative, we are updating the /import page to suggest that users use our importing guided flows (PR here: https://github.com/Automattic/jetpack/pull/42693).
This PR ensures the user can return to the original page when they start the flow from there, regardless of whether they continue on the site setup or continue the site migration flow.


## Testing Instructions
* Create a free site
* Go to `/setup/site-setup/importList?siteSlug={YOUR_SITE_SLUG}&ref=wp-admin-importers-list`  
* Click "WordPress"
* Choose the Import option
* Click on the UI back button and check if you are on the import vs migrate again
* Click on the UI back button to check if you are on the importList again
* Click on the UI back button to check if you are on the /import.php page again. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

gabrielcaires committed:
Use flow entry point to manage back button edgy cases

gabrielcaires committed:
Initialize entry point

gabrielcaires commented on client/landing/stepper/declarative-flow/flows/site-setup-flow/site-setup-flow.ts:
```diff
@@ -746,6 +759,14 @@ const siteSetupFlow: FlowV1 = {
 			skippedCheckout,
 		] );
+
+		const { get, set } = useFlowState();
```
This change is part of the useSideEffect() hook (Github way is not making it clear)

daledupreez commented on client/landing/stepper/declarative-flow/flows/site-setup-flow/site-setup-flow.ts:
```diff
@@ -272,6 +275,9 @@ const siteSetupFlow: FlowV1 = {
 		useRedirectDesignSetupOldSlug( currentStep, navigate );
+		const { get } = useFlowState();
+		const entryPoint = get( 'flow' )?.entryPoint;
+		const siteAdminUrl = useSelector( ( state ) => getSiteAdminUrl( state, siteId ) );
```
I see that `adminUrl` is defined above. Is there a reason we need a second selector here?

gabrielcaires replied:
Thanks, removed here [78bd9f4](https://github.com/Automattic/wp-calypso/pull/102224/commits/78bd9f4191b5c19b5aea1d9490a06eed88381b67)

sixhours commented:
I don't see the Migrate vs Import step when I go through the described testing instructions with this branch. Is that intentional?

https://github.com/user-attachments/assets/b5544957-d13b-4848-b188-a08e149d40ba

Minor note for the testing instructions, don't forget to add the `&` before the `ref` in the query param.

gabrielcaires committed:
Add missing entrypoint

gabrielcaires committed:
Rename param

gabrielcaires commented:
> I don't see the Migrate vs Import step when I go through the described testing instructions with this branch. Is that intentional?

Fixed [3540f96](https://github.com/Automattic/wp-calypso/pull/102224/commits/3540f96258553078b5803d0e001311928bf76b0f)

Sorry about that, I forget to push 🤦 

> Minor note for the testing instructions, don't forget to add the & before the ref in the query param.

Thanks, I updated the docs

gabrielcaires committed:
Remove uncessary const

gabrielcaires committed:
Remove duplicated selectors

sixhours commented:
I think we're losing the `ref` at some point along the way, because when I click Back from the Migrate vs. Import step now, I'm taken to the Identify step rather than the importerList:

https://github.com/user-attachments/assets/9a00c7cb-084e-4ded-a2e1-4e017b5eaaad

It behaves correctly when I manually re-add the `ref=wp-admin-importers-list` param.

gabrielcaires committed:
Add missing ref param

gabrielcaires commented:
Thanks,

I fixed it here [735b71f](https://github.com/Automattic/wp-calypso/pull/102224/commits/735b71f69d982e3a5ff18e80026a8c6ca7db389c)

https://github.com/user-attachments/assets/7a3c24ee-f025-46fe-9001-90d3baab7c72

Sorry about the back and forward; it is pretty complicated to manage, and our test suite is useless here :(

sixhours reviewed (approved):
LGTM!

alshakero commented:
I'm really not a big fan of this 😢 

It's tying two+ flows in untraceable ways. It would be great if you can explore removing `goBack` from the flow and making [this condition smarter](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/hooks/use-step-navigation-with-tracking/index.ts#L54). 

**My take would be**

1. Remove `goBack` function entirely from `site-setup`.
2. Explore checking if `document.referrer === config('hostname')` as a condition in `canUserGoBack`.
3. Remove [this line](https://github.com/Automattic/wp-calypso/blob/a39c421c0081b80f198e7f64a98b6ed41926b4fa/client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/unified-design-picker.tsx#L661) because it messes with the URL for no reason.

This will outsource the task to the browser and align the back button of the flow with the one from the browser. 

Note that I haven't tried this, but I suspect it may work with much less complexity.

gabrielcaires commented:
> I'm really not a big fan of this 😢
> 
I hated this solution 😠 

and I don't think it is the final solution.

The big issue regarding the `history` usage is that both flows are using customized GoBack, and we can't use any history strategy if we have at least one customized GoBack because it messes up the flow.

I think the ideal solution is site migration and site setup + adopting the new callback @daledupreez is suggesting [here](https://github.com/Automattic/wp-calypso/pull/101671)  so one flow can point directly to another using the window. `location.replace` without messing with history object. 

The document.referrer also doesn't work because it only returns the origin hostname, and we have multiple points where this flow can be triggered. 

Docs about https://developer.chrome.com/blog/referrer-policy-new-chrome-default

gabrielcaires commented:
Merging to unblock the code blue issue
